xds/internal/server/listener_wrapper_test.go

```diff
@@ -21,154 +21,226 @@ package server
+	lis, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatalf("Failed to create a local TCP listener: %v", err)
 	}
 
+	host, port := hostPortFromListener(t, lis)
+	lisResourceName := fmt.Sprintf(e2e.ServerListenerResourceNameTemplate, net.JoinHostPort(host, strconv.Itoa(int(port))))
+	params := ListenerWrapperParams{
+		Listener:             lis,
+		ListenerResourceName: lisResourceName,
+		XDSClient:            xdsC,
+	}
+	l, readyCh := NewListenerWrapper(params)
+	if l == nil {
+		t.Fatalf("NewListenerWrapper(%+v) returned nil", params)
+	}
+	t.Cleanup(func() { l.Close() })
+	return readyCh, host, port, lisResourceName
 }
 
-type s struct {
-	grpctest.Tester
+// Tests the case where a listenerWrapper is created and following happens:
+//
+//   - the management server returns a Listener resource that is NACKed. Test
+//     verifies that the listenerWrapper does not become ready.
+//   - the management server returns a Listener resource that does not match the
+//     address to which our net.Listener is bound to. Test verifies that the
+//     listenerWrapper does not become ready.
+//   - the management server returns a Listener resource that contains an inline
+//     Route Configuration. Test verifies that the listenerWrapper becomes ready.
+func (s) TestListenerWrapper_InlineRouteConfig(t *testing.T) {
+	mgmtServer, nodeID, ldsResourceNamesCh, _, xdsC := xdsSetupFoTests(t)
+	readyCh, host, port, lisResourceName := createListenerWrapper(t, xdsC)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	waitForResourceNames(ctx, t, ldsResourceNamesCh, []string{lisResourceName})
+
+	// Configure the management server with a listener resource that is expected
+	// to be NACKed.
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{badListenerResource(t, lisResourceName)},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Verify that there is no message on the ready channel.
+	sCtx, sCtxCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+	defer sCtxCancel()
+	select {
+	case <-readyCh:
+		t.Fatalf("Ready channel written to after receipt of a bad Listener update")
+	case <-sCtx.Done():
+	}
+
+	// Configure the management server with a listener resource that does not
+	// match the address to which our listener is bound to.
+	resources.Listeners = []*v3listenerpb.Listener{e2e.DefaultServerListener(host, port+1, e2e.SecurityLevelNone, route1)}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Verify that there is no message on the ready channel.
+	select {
+	case <-readyCh:
+		t.Fatalf("Ready channel written to after receipt of a bad Listener update")
+	case <-sCtx.Done():
+	}
+
+	// Configure the management server with a good update that does not contain
```

Comment:
Nit here and top level: can you explain what makes this update good? (i.e. matching host and port of listener).